### PR TITLE
Added 'Allow Empty' checkbox to commit form, and updated command logic

### PR DESCRIPTION
**Pull request by darkcamper**
Implements #4631

## Proposed changes

-  Add new option to the commit dialog to allow commits with no changes.

### Before

before-4631

### After

after-4631

## Test methodology 
Manual testing

## Test environment(s) 
- GIT 2.29
- Windows 10, 20H2

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

**Comment by pmiossec** [comment]
We need your update on the `contributors.txt` file, if not done previously, preferably in another pull request.

And you need to update the translation: https://github.com/gitextensions/gitextensions/wiki/Translations#v34-and-newer

**Comment by darkcamper** [comment]
Ok! I've updated this PR with the updated translation file and I'll open another PR for the contributors file

File: GitUI/Translation/English.xlf [hunk 1]
```diff
@@ -3201,6 +3201,10 @@ gitex.cmd / gitex (located in the same folder as GitExtensions.exe):</source>
         <source>Commit</source>
         <target />
       </trans-unit>
+      <trans-unit id="AllowEmpty.Text">
+        <source>Allow &amp;Empty</source>
```
Comment (mstv) [inline GitUI/Translation/English.xlf:3205]:
```suggestion
        <source>Allow &amp;empty</source>
```
Could you also align the `&Amend commit` checkbox to our capitalization convention, please?

**Review by mstv — COMMENTED** [review]
Please revert any unnecessary changes to `FormCommit.Designer.cs`. Or else create a separate commit with how the designer saves the file without UI changes. Then add the new checkbox.
Have you adapted the tab order?
Force-pushing this feature branch is welcome.

The check state of the `Allow empty` must affect the `CommitAndPush` button text in a similar way as `Amend commit` does.

**Comment by RussKie** [comment]
This looks to me as a rather narrow use case, and with that I oppose to the
checkbox being added to the main screen.

It can probably be added to the dropdown menu that hosts additional options
(e.g. "refresh on focus", "keep open", etc.)

**Comment by drewnoakes** [comment]
This adds too much complexity to the UI in my opinion. I prefer allowing the user to commit an empty message without the checkbox. Perhaps a dialog box is shown saying "No commit message has been entered. Commit anyway?" With ok/cancel.

**Comment by RussKie** [comment]
I could agree to this.

**Comment by darkcamper** [comment]
@drewnoakes @RussKie, just to clarify, this feature would allow you to make a commit with no changes (no file additions/modifications), but the message would still be mandatory.

I think the dialog box would work really well when there are no changes in the working directory:

`'There are no files staged for this commit' (Ok) => 'There are no files staged for this commit. Commit anyway?' (Yes/No)`

But when there are pending changes (not staged), there is already a dialog asking if you want to stage everything, and commit.

`
'There are no files staged for this commit. Stage and commit all unstaged files?' (Yes/No)`

One possibility would be adding a "Cancel" option

```
'There are no files staged for this commit. Stage and commit all unstaged files?' (Yes/No/Cancel)
--> Yes: Old 'yes' behavior
--> No: 'There are no files staged for this commit. Commit anyway? (Yes/No)' dialog
--> Cancel: Old 'no' behavior
```

Although I'm not sure if it's a bit confusing. What do you think?

**Comment by pmiossec** [comment]
Perhaps you could use a task dialog (like in https://github.com/gitextensions/gitextensions/pull/8563 ) when their is no files staged with 2 possible actions "Stage all files and commit"(disabled if no files changed) and "Create an empty commit". And indeed a cancel button.

**Comment by RussKie** [comment]
That could work

**Comment by darkcamper** [comment]
It seems like a better idea. I'm giving it a try

**Comment by darkcamper** [comment]
taskdialog-emptycommit_disablestage

Well, this is the result. Anything you'd change?

**Comment by pmiossec** [comment]
Was there an icon in the message box you replaced?
If yes, I would display the same...
If not, I propose (wait for other advices...) to add the warning icon.

**Comment by darkcamper** [comment]
There was a question mark icon, but it's not available in TaskDialog (and it seems that Microsoft is advising against using question icons outside 'Help entry points'...)  

If having an icon is preferred I'd use the warning icon too

**Comment by drewnoakes** [comment]
I like the improvements here 👍 

Please also verify that the <kbd>Esc</kbd> key is equivalent to cancelling the dialog. (I think that is the default behaviour with `TaskDialog`, but please check.)

File: GitUI/CommandsDialogs/FormCommit.Designer.cs [hunk 2]
```diff
@@ -23,61 +23,62 @@ private void InitializeComponent()
         {
```
Comment (RussKie) [inline GitUI/CommandsDialogs/FormCommit.Designer.cs:23]: The designer is evil... 
Please undo _all_ unrelated changed.

**Review by RussKie — COMMENTED** [review]
> taskdialog-emptycommit_disablestage
> 
> Well, this is the result. Anything you'd change?

Nice!

I'd like to ask for a few changes:
1. Please remove the trailing dots in the command links.
2. Use the same icon as in the dialog below.
3. I'd change the heading to something like "Confirm commit"

E.g. 
image

**Comment by darkcamper** [comment]
@RussKie Ok to those changes! (they'll certainly make the new dialog more consistent with the rest of the application). I'll also make sure that all the mess that the winforms designer introduced is cleaned up.

**Comment by darkcamper** [comment]
While doing some tests I came up with a question about the "Push/Commit & Push" button. Should this new feature affect in any way that button, when there aren't unstaged changes? (the case when there are unstaged changes is already covered by the new dialog)

Previously, when you couldn't make empty commits, it made sense for the button to only allow "pushing to remote" in that situation. But now, I wonder if we should ask the user if they want to "make an empty commit & push/push" or just leave that button as is.

**Comment by gerhardol** [comment]
> 
> 
> While doing some tests I came up with a question about the "Push/Commit & Push" button. Should this new feature affect in any way that button, when there aren't unstaged changes? (the case when there are unstaged changes is already covered by the new dialog)
> 
> Previously, when you couldn't make empty commits, it made sense for the button to only allow "pushing to remote" in that situation. But now, I wonder if we should ask the user if they want to "make an empty commit & push/push" or just leave that button as is.

A special case to a special case, not so important to me.
If you bother, "make an empty commit & push/push" seems more complete to me.

**Comment by RussKie** [comment]
How would it look like in English and other languages we support?

**Comment by darkcamper** [comment]
After some consideration, I think that overloading the 'Push/Commit & Push' button with yet another option could be confusing and would bring little benefit. After all, If you want to create & push an empty commit just can just use the 'Commit' button, choose the 'Make empty commit' in the new dialog, and then press the 'Push' button.

Besides I can't really think of a good caption/flow for that button, if we finally try to squeeze the 'make empty commit' into it.

Naturally, if anyone cames up with a good idea, I'd be willing to make the necessary changes.

**Comment by darkcamper** [comment]
I've force pushed all the changes we've talked about (well, I did it a week ago but forgot to post a comment 😓).

**Review by RussKie — APPROVED** [review]

**Comment by RussKie** [comment]
Thank you

**Comment by darkcamper** [comment]
Thank you all for your help with my first contribution!

**Pull request merged**